keymaster: add keymint support

massonju opened the pull request:
In order to facilitate the support of keymint, all the logics/code in the keymaster service has been realigned with trusty first.

Indeed there is nothing specific for OP-TEE, compared to trusty the main difference are:
- forward operations from IPC to the Trusted Application
- support crypto operations in the Trusted Application

Thus all the code related to the keymaster service is now realigned with trusty: `system/core/trusty/keymaster`

We follow the exact same tree structure and naming convention.

In that way all the futur fixes/improvments made in trusty service can be applied easily for OP-TEE keymaster.

All the crypto operations not yet supported should be reported by the Trusted Application.

Closes #31

jockebech commented:
Hi @massonju , I've started review this, but since it's quite a lot of code, it'll probably take some time. But just as you know, someone is looking at this. Having said, that since this IIUIC is mostly a code from Trusty, the code should already have been reviewed at Google.

massonju commented:
Hi @jbech-linaro

Thanks for the feedbacks and yes sorry for that, indeed that is a lot of code changed and I don't really know how to reduce/split that since it realign the whole keymaster services code to trusty ...

> since it's quite a lot of code, it'll probably take some time

Yes I totally understand, no problem

FYI we have a fork here:
https://gitlab.baylibre.com/baylibre/ti/android/aosp/external/kmgk

And we aims to contribute back steps by steps, it contains several fixes that you might be interested (bugs, CTS/VTS ...)

Thanks

jockebech commented on keymaster/3.0/OpteeKeymaster3Device.cpp:
```diff
@@ -0,0 +1,464 @@
+Return<void> OpteeKeymaster3Device::getHardwareFeatures(getHardwareFeatures_cb _hidl_cb) {
+    _hidl_cb(true /* is_secure */, true /* supports_ec */,
+             true /* supports_symmetric_cryptography */, true /* supports_attestation */,
+             true /* supportsAllDigests */, "OpteeKeymaster", "OP-TEE Foundation");
```
There is no such thing as OP-TEE Foundation. It's either simply the OP-TEE project or OP-TEE by TrustedFirmware.

massonju replied:
Got it I will change this, I followed the current infos:
https://github.com/linaro-swg/kmgk/blob/master/keymaster/3.0/optee_keymaster3_device.cpp#L247

jockebech commented on keymaster/3.0/OpteeKeymaster3Device.cpp:
```diff
@@ -0,0 +1,464 @@
+
+    DeleteKeyResponse response(impl_->message_version());
+    impl_->DeleteKey(request, &response);
+
```
Do you know why we don't check `response error` here like done in the previous functions? Here we just return with the legacy conversion based on the error code.(Same for the `deleteAllKeys()` below). I've noted that the other returns `Void`, so there is a difference as such.

massonju replied:
In `deleteKey` we let the caller handle the error (that's why we return `Return<ErrorCode>` I guess):
Example with vold:
https://android.googlesource.com/platform/system/vold/+/refs/tags/android-14.0.0_r1/Keystore.cpp#193

jockebech commented on keymaster/3.0/OpteeKeymaster3Device.cpp:
```diff
@@ -0,0 +1,464 @@
+}
+
+Return<ErrorCode> OpteeKeymaster3Device::destroyAttestationIds() {
+    return ErrorCode::UNIMPLEMENTED;
```
I can see that [this is an optional feature](https://source.android.com/docs/security/features/keystore/implementer-ref?hl=en#destroy_attestation_ids), but they also say "but highly recommended". Is this something worth looking into? If not this patch set, perhaps future patch sets.?

massonju replied:
Yes maybe we can look at it in future patch, that is a good point.

For now I just followed trusty:
https://android.googlesource.com/platform/system/core/+/refs/tags/android-14.0.0_r15/trusty/keymaster/3.0/TrustyKeymaster3Device.cpp#371

jockebech commented on keymaster/3.0/OpteeKeymaster3Device.cpp:
```diff
@@ -0,0 +1,464 @@
+        response.error = KM_ERROR_INVALID_INPUT_LENGTH;
+    } else {
+        if (ser_size + inp_size > OPTEE_KEYMASTER_SEND_BUF_SIZE) {
+            inp_size = OPTEE_KEYMASTER_SEND_BUF_SIZE - ser_size;
```
I have a little bit of trouble following this one. If they size isn't big enough, we just make the input size smaller? This implies that `inp_size` becomes smaller than `input.size()`. Aren't we going to miss data in this case?

massonju replied:
Well I agree I doesn't make sense to me to do the operation with an input shrink, the final result won't be correct ...

I will suggest to return `KM_ERROR_INVALID_INPUT_LENGTH` ?

jockebech commented on keymaster/Android.bp:
```diff
@@ -0,0 +1,116 @@
+    cflags: [
+        "-DANDROID_BUILD",
+        "-Wall",
+        "-Wextra",
```
Would it make sense to add the additional flags as done for the `wait_for_keymaster_optee` below?

massonju replied:
Humm maybe but it should be done another patch.

The intention of this commit is to do a pure conversion of Android.mk to Android.bp without changing the current parameters (shared libs, cflags ...)

jockebech commented on keymaster/Android.bp:
```diff
@@ -0,0 +1,116 @@
+    vendor: true,
+    cflags: [
+        "-Wall",
+        "-Wextra",
```
Same here.

jockebech commented on keymaster/OpteeKeymaster.cpp:
```diff
@@ -84,26 +76,35 @@ int OpteeKeymaster::Initialize() {
+    ConfigureVendorPatchlevelResponse vendor_rsp = ConfigureVendorPatchlevel(vendor_req);
+    if (vendor_rsp.error != KM_ERROR_OK) {
+        LOG(ERROR) << "Failed to configure keymaster vendor patchlevel: " << vendor_rsp.error;
+        // Don't fail if this message isn't understood.
```
Could you explain more in detail (as a comment in the code), why this is OK, i.e., to not return any error if this happens.

massonju replied:
No idea, it's copy-paste from trusty ...

In our fork we let the TA handle this request:
https://gitlab.baylibre.com/baylibre/ti/android/aosp/external/kmgk/-/blob/ti-android-14/keymaster/ta/keystore_ta.c?ref_type=heads#L1819

The function was stubbed which means we return `KM_ERROR_OK` but we do nothing.

So I think either we do this kind of logic or we remove this code ?

jockebech commented on keymaster/include/optee_keymaster/ipc/keymaster_ipc.h:
```diff
@@ -18,7 +18,7 @@
 
-#define KEYMASTER_PORT "com.android.trusty.keymaster"
+#define KEYMASTER_PORT "com.android.optee.keymaster"
 #define KEYMASTER_MAX_BUFFER_LENGTH 4096
```
Is this related to `PAGE_SIZE` in `keymaster/include/optee_keymaster/ipc/optee_keymaster_ipc.h`? If yes, should this define be used instead of PAGE_SIZE?

massonju replied:
Well actually this macro is never used and was imported from trusty.

In trusty it's used here:
https://android.googlesource.com/platform/system/core/+/refs/tags/android-14.0.0_r15/trusty/keymaster/ipc/trusty_keymaster_ipc.cpp#140

But this part is specific to trusty (handle communication with trusty OS).

So I would say this macro can be removed

jockebech commented on keymaster/include/optee_keymaster/ipc/keymaster_ipc.h:
```diff
@@ -53,7 +53,31 @@ enum keymaster_command : uint32_t {
+    KM_GET_HW_INFO                  = (35 << KEYMASTER_REQ_SHIFT),
+    KM_GENERATE_CSR_V2              = (36 << KEYMASTER_REQ_SHIFT),
+
+    // Bootloader/provisioning calls.
+    KM_SET_BOOT_PARAMS = (0x1000 << KEYMASTER_REQ_SHIFT),
+    KM_SET_ATTESTATION_KEY = (0x2000 << KEYMASTER_REQ_SHIFT),
+    KM_APPEND_ATTESTATION_CERT_CHAIN = (0x3000 << KEYMASTER_REQ_SHIFT),
+    KM_ATAP_GET_CA_REQUEST = (0x4000 << KEYMASTER_REQ_SHIFT),
+    KM_ATAP_SET_CA_RESPONSE_BEGIN = (0x5000 << KEYMASTER_REQ_SHIFT),
+    KM_ATAP_SET_CA_RESPONSE_UPDATE = (0x6000 << KEYMASTER_REQ_SHIFT),
+    KM_ATAP_SET_CA_RESPONSE_FINISH = (0x7000 << KEYMASTER_REQ_SHIFT),
+    KM_ATAP_READ_UUID = (0x8000 << KEYMASTER_REQ_SHIFT),
+    KM_SET_PRODUCT_ID = (0x9000 << KEYMASTER_REQ_SHIFT),
+    KM_CLEAR_ATTESTATION_CERT_CHAIN = (0xa000 << KEYMASTER_REQ_SHIFT),
+    KM_SET_WRAPPED_ATTESTATION_KEY = (0xb000 << KEYMASTER_REQ_SHIFT),
+    KM_SET_ATTESTATION_IDS = (0xc000 << KEYMASTER_REQ_SHIFT),
+    KM_SET_ATTESTATION_IDS_KM3 = (0xc001 << KEYMASTER_REQ_SHIFT),
+    KM_CONFIGURE_BOOT_PATCHLEVEL = (0xd000 << KEYMASTER_REQ_SHIFT),
```
Since the above is aligned, it would probably make sense to align this the same.

massonju replied:
yes agree, initial alignment from trusty

jockebech commented on keymaster/include/optee_keymaster/ipc/optee_keymaster_ipc.h:
```diff
@@ -14,29 +14,28 @@
 
 const uint32_t OPTEE_KEYMASTER_RECV_BUF_SIZE = 2 * PAGE_SIZE;
-const uint32_t OPTEE_KEYMASTER_SEND_BUF_SIZE = 2 * PAGE_SIZE;
+const uint32_t OPTEE_KEYMASTER_SEND_BUF_SIZE =
```
This seems to be a bit smaller compared to the old version, is that intentional and OK?

massonju replied:
humm agree I don't see a reason to reduce the size, the `tipc header` seems to be a trusty thing ...

I prefer to keep the old version

jockebech commented on keymaster/ipc/optee_keymaster_ipc.cpp:
```diff
@@ -54,6 +44,12 @@ int optee_keymaster_connect(void) {
+    res = TEEC_InitializeContext(NULL, &ctx);
+    if (res != TEEC_SUCCESS) {
+        ALOGE("TEEC_InitializeContext failed with code 0x%x", res);
+        return (int)res;
```
`TEEC_Result` is an `uint32_t` (where the upper bits are used), so casting to an int is probably not a good idea. I can see that this was like that in the old code as well. But I think it would be better to let this function return uint32_t if possible. Otherwise we should transform the return values to some other values.

massonju replied:
yes I agree and I have already future changes for that but the goal of the commit is to "just" realign the code with trusty.

So I did not want (yet) change this part and keep the current code:
https://github.com/linaro-swg/kmgk/blob/master/keymaster/ipc/optee_keymaster_ipc.cpp#L41

jockebech commented on keymaster/keymint/OpteeKeyMintDevice.cpp:
```diff
@@ -0,0 +1,349 @@
+
+ScopedAStatus OpteeKeyMintDevice::getKeyCharacteristics(
+        const vector<uint8_t>& keyBlob,
+        const vector<uint8_t>& clientId,  //
```
Remove `//`

jockebech commented on keymaster/keymint/OpteeKeyMintDevice.cpp:
```diff
@@ -0,0 +1,349 @@
+ScopedAStatus OpteeKeyMintDevice::getKeyCharacteristics(
+        const vector<uint8_t>& keyBlob,
+        const vector<uint8_t>& clientId,  //
+        const vector<uint8_t>& appData,   //
```
Remove `//`

jockebech commented on keymaster/keymint/OpteeKeyMintDevice.cpp:
```diff
@@ -0,0 +1,349 @@
+}
+
+ScopedAStatus OpteeKeyMintDevice::importWrappedKey(const vector<uint8_t>& wrappedKeyData,
+                                                    const vector<uint8_t>& wrappingKeyBlob,  //
```
Remove `//`

jockebech commented on keymaster/keymint/OpteeKeyMintDevice.cpp:
```diff
@@ -0,0 +1,349 @@
+                                                    const vector<uint8_t>& wrappingKeyBlob,  //
+                                                    const vector<uint8_t>& maskingKey,
+                                                    const vector<KeyParameter>& unwrappingParams,
+                                                    int64_t passwordSid,  //
```
Remove `//`

jockebech commented on keymaster/keymint/OpteeKeyMintDevice.cpp:
```diff
@@ -0,0 +1,349 @@
+    request.SetWrappingMaterial(wrappingKeyBlob.data(), wrappingKeyBlob.size());
+    request.SetMaskingKeyMaterial(maskingKey.data(), maskingKey.size());
+    request.additional_params.Reinitialize(KmParamSet(unwrappingParams));
+    request.password_sid = static_cast<uint64_t>(passwordSid);
```
Since we cast this (and next line) to `uint64_t`, could we and would it make sense to change the function parameters to they are the same?

massonju replied:
I cannot change the type, it's defined by the vndk:
https://android.googlesource.com/platform/prebuilts/vndk/v33/+/refs/tags/android-14.0.0_r15/arm64/include/generated-headers/hardware/interfaces/security/keymint/aidl/android.hardware.security.keymint-V2-ndk-source/gen/include/aidl/android/hardware/security/keymint/IKeyMintDevice.h#66

jockebech commented on keymaster/keymint/OpteeKeyMintDevice.cpp:
```diff
@@ -0,0 +1,349 @@
+}
+
+ScopedAStatus OpteeKeyMintDevice::deleteAllKeys() {
+    // There's nothing to be done to delete software key blobs.
```
Perhaps better to say that there are no specific parameters needed and no need to set key material when deleting all keys. I.e., a request and response is sufficient.

jockebech commented on keymaster/keymint/OpteeKeyMintOperation.cpp:
```diff
@@ -0,0 +1,184 @@
+
+ScopedAStatus OpteeKeyMintOperation::updateAad(
+        const vector<uint8_t>& input, const optional<HardwareAuthToken>& authToken,
+        const optional<TimeStampToken>& /* timestampToken */) {
```
Comment superfluous.

jockebech commented on keymaster/keymint/OpteeKeyMintOperation.cpp:
```diff
@@ -0,0 +1,184 @@
+
+ScopedAStatus OpteeKeyMintOperation::update(const vector<uint8_t>& input,
+                                             const optional<HardwareAuthToken>& authToken,
+                                             const optional<TimeStampToken>& /* timestampToken */,
```
Comment superfluous.

jockebech commented on keymaster/keymint/OpteeKeyMintOperation.cpp:
```diff
@@ -0,0 +1,184 @@
+ScopedAStatus OpteeKeyMintOperation::finish(const optional<vector<uint8_t>>& input,      //
+                                             const optional<vector<uint8_t>>& signature,  //
+                                             const optional<HardwareAuthToken>& authToken,
+                                             const optional<TimeStampToken>& /* timestampToken */,
```
Comment superfluous.

jockebech commented on keymaster/keymint/OpteeKeyMintOperation.cpp:
```diff
@@ -0,0 +1,184 @@
+    return ScopedAStatus::ok();
+}
+
+ScopedAStatus OpteeKeyMintOperation::finish(const optional<vector<uint8_t>>& input,      //
+                                             const optional<vector<uint8_t>>& signature,  //
```
Remove `//`

jockebech commented on keymaster/keymint/OpteeSharedSecret.cpp:
```diff
@@ -0,0 +1,57 @@
+    for (size_t i = 0; i < params.size(); ++i) {
+        request.params_array.params_array[i].seed = {params[i].seed.data(), params[i].seed.size()};
+        if (sizeof(request.params_array.params_array[i].nonce) != params[i].nonce.size()) {
+            return kmError2ScopedAStatus(KM_ERROR_INVALID_ARGUMENT);
```
Don't we need to delete `request.params_array.params_array` (allocated by `new` at line 41) before returning here, if we have an error? I.e., to prevent memory leak?

jbech-linaro reviewed (changes requested):
Mostly minor things, but please have a look at my comments.

massonju commented:
@jbech-linaro  thanks for the review and I agree with all the comments.

I think we need to be agree about what strategy to do when we integrate/copy code from trusty.

Right now I blindly copy/paste the code from trusty and do several `sed` /`rename` of `trusty` -> `optee`.
Ok great we are aligned with trusty and all the futures updates will be applied easily.
And this is interesting because trusty code (only services part) follow updates from keymaster also.

**But** we also get some code where we might not be interested  or agreed (your comments).

How do you think we should handle this ?

1. Copy paste trusty code, `sed`/`rename`,  no modifications of Trusty code (except renaming), focus only the TA
OR
2. Copy paste trusty code, `sed`/`rename`, do a review of the Trusty code imported following our needs
OR
3. Something else ...

I'm okay with any options, just want to make sure we are on the same page

jockebech commented:
@massonju , for this particular PR and since the OP-TEE KM/GK implementation hasn't been maintained, I'm fine with doing this in two stages, first copy the code without any changes and then after that fix the issues seen.

I guess the "fix the issues" is the major question you had on mind, i.e., how should that be done. I think if we just blindly copy paste everything but TA code, then we're making a shortcut that eventually could have negative outcome for the overall security. I.e., what we get in convenience we eventually have to pay with the overall security.

I.e., this is source code for security features, so it's on us to try to make our best to ensure that this as good as possible. Therefore I want review the code coming from Google when doing rebases etc. If there are issues I see two ways forward: 1) Fix the issues in our tree (a must) 2) Preferably let Google know about the fixes, so they can decide whether they agree and are willing to fix them or not.

How does this sound to you? If, OK then we can merge this as is and then, you can send a follow up PR addressing review comments from this PR?.

massonju commented:
@jbech-linaro
okay yes that makes sense to me

> If, OK then we can merge this as is and then, you can send a follow up PR addressing review comments from this PR?.

I would prefer to add commits "fix the issues" in this PR on top of the trusty "rebase/realign" commit.
Thus we can track the fixes in response to your comments and we merge everything together

massonju commented:
@jbech-linaro sorry for the late reply, I've been quite busy last days ...

I've addressed most of your comments and you will see that there is two fixup commits.
I think we can apply them directly in their corresponding commit but I would like your opinion on that:
1 commit: pure copy/rename from trusty with some clean up
VS
2 commits: 1 commit pure copy/rename from trusty, 1 commit clean up

jockebech commented:
> I think we can apply them directly in their corresponding commit but I would like your opinion on that:
> 1 commit: pure copy/rename from trusty with some clean up
> VS
> 2 commits: 1 commit pure copy/rename from trusty, 1 commit clean up

I think alternative two sounds better, since that makes it possible to track the changes a bit easier.

syan10 commented:
> In order to facilitate the support of keymint, all the logics/code in the keymaster service has been realigned with trusty first.
> 
> Indeed there is nothing specific for OP-TEE, compared to trusty the main difference are:
> 
> * forward operations from IPC to the Trusted Application
> * support crypto operations in the Trusted Application
> 
> Thus all the code related to the keymaster service is now realigned with trusty: `system/core/trusty/keymaster`
> 
> We follow the exact same tree structure and naming convention.
> 
> In that way all the futur fixes/improvments made in trusty service can be applied easily for OP-TEE keymaster.
> 
> All the crypto operations not yet supported should be reported by the Trusted Application.
> 
> Closes #31

hi @massonju,  
Could i know when does the Trust Application is ready for this feature? Thanks

massonju commented:
Hi @syan10 

> Could i know when does the Trust Application is ready for this feature? Thanks

With this PR Keymint can run without any additional changes in the Trust Application.
However several crypto operations are not yet supported:
https://github.com/linaro-swg/kmgk/blob/master/keymaster/ta/keystore_ta.c#L1958

If you run CTS/VTS you will see what operations are not yet supported.

jockebech commented:
@massonju feel free to add my RB-tag to the patches and then I'll merge this:
`Reviewed-by: Joakim Bech <joakim.bech@linaro.org>`

syan10 commented:
> Hi @syan10
> 
> > Could i know when does the Trust Application is ready for this feature? Thanks
> 
> With this PR Keymint can run without any additional changes in the Trust Application. However several crypto operations are not yet supported: https://github.com/linaro-swg/kmgk/blob/master/keymaster/ta/keystore_ta.c#L1958
> 
> If you run CTS/VTS you will see what operations are not yet supported.

Thanks @massonju.

How about the RKP support plan? seems it doesn't support yet?
